internal/engine/compiler/engine.go

```diff
@@ -455,15 +464,21 @@ func (e *engine) CompileModule(ctx context.Context, module *wasm.Module) error {
 
 	importedFuncs := module.ImportFuncCount()
 	funcs := make([]*code, len(module.FunctionSection))
+	ln := len(listeners)
 	for i, ir := range irs {
+		var withListener bool
+		if i < ln {
+			withListener = listeners[i] != nil
+		}
+
```

Comment:
this makes it possible for us to assemble machine code on the function listeners calling only when a function has non-nil listeners, and in the machine code generation, we call the listener on preamble and function return as opposed to the case of interpreter where it's calling it on the caller side.